testplan/runnable/interactive/http.py

```diff
@@ -196,37 +209,53 @@ def put(self, test_uid):
                     raise werkzeug.exceptions.NotFound
 
                 try:
-                    new_test = TestGroupReport.shallow_deserialize(
+                    new_test = _deserialize_report_entry(
                         flask.request.json, current_test
                     )
                 except marshmallow.exceptions.ValidationError as e:
                     raise werkzeug.exceptions.BadRequest(str(e))
 
                 _check_uids_match(current_test.uid, new_test.uid)
+                new_runtime_status = flask.request.json.get("runtime_status")
 
                 # Trigger a side-effect if either the report or environment
                 # statuses have been updated.
-                if _should_run(current_test.runtime_status):
-                    if current_test.env_status != new_test.env_status:
-                        raise werkzeug.exceptions.BadRequest(
-                            "env_status cannot change when test status is "
-                            "changing. Current env status is {}".format(
-                                current_test.env_status
-                            )
-                        )
-
+                if _should_reset(
+                    current_test.uid,
+                    current_test.runtime_status,
+                    new_runtime_status,
+                ):
+                    ihandler.reset_test(test_uid, await_results=False)
+                elif _should_run(
+                    current_test.uid,
+                    current_test.runtime_status,
+                    new_runtime_status,
+                ):
+                    _check_env_status_match(
+                        current_test.env_status, new_test.env_status
+                    )
                     _check_execution_order(ihandler.report, test_uid=test_uid)
-                    new_test.runtime_status = RuntimeStatus.RUNNING
                     ihandler.run_test(test_uid, await_results=False)
                 else:
-                    env_action = self._check_env_transition(
+                    if current_test.runtime_status in (
```

Comment:
An error message will shows in the center pane. Actually this will not happen on UI, because when testcase is running or waiting you can not click the play/rerun button, it is disabled. However, we can send request with API, thus the server side will know that it's currently running test we cannot accept other request (env changing)